### PR TITLE
Feat add Amazon Bedrock support

### PR DESCRIPTION
**Pull request by usamimeri**
## **User description**
**Features**
#1174 
- add support for Amazon Bedrock Provider (Meta, Mistral, Amazon, Ai21, Anthropic, Cohere)
- add test for Amazon Bedrock

**config**
```yaml
llm:
  api_type: "bedrock" 
  model: "meta.llama3-70b-instruct-v1:0"  # see metagpt/provider/bedrock/utils.py
  region_name: "REGION"
  access_key: "YOUR_ACCESS_KEY"
  secret_key: "YOUR_SECRET_KEY"
```
**test result**
```============================================================================== test session starts ===============================================================================
platform win32 -- Python 3.9.19, pytest-8.1.1, pluggy-1.5.0
rootdir: D:\python_projects\MetaGPT
plugins: anyio-4.3.0, asyncio-0.23.6, cov-5.0.0, html-4.1.1, metadata-3.1.1, mock-3.14.0, timeout-2.3.1, xdist-3.5.0
asyncio: mode=strict
collected 129 items

tests\metagpt\provider\postprocess\test_base_postprocess_plugin.py .                                                                                                        [  0%]
tests\metagpt\provider\postprocess\test_llm_output_postprocess.py .                                                                                                         [  1%]
tests\metagpt\provider\test_amazon_bedrock_api.py ............................................................................................                              [ 72%]
tests\metagpt\provider\test_anthropic_api.py .                                                                                                                              [ 73%]
tests\metagpt\provider\test_azure_llm.py .                                                                                                                                  [ 74%]
tests\metagpt\provider\test_base_llm.py ..                                                                                                                                  [ 75%]
tests\metagpt\provider\test_dashscope_api.py .                                                                                                                              [ 76%]
tests\metagpt\provider\test_general_api_base.py .......                                                                                                                     [ 82%]
tests\metagpt\provider\test_general_api_requestor.py ...                                                                                                                    [ 84%]
tests\metagpt\provider\test_google_gemini_api.py .                                                                                                                          [ 85%]
tests\metagpt\provider\test_human_provider.py .                                                                                                                             [ 86%]
tests\metagpt\provider\test_metagpt_llm.py .                                                                                                                                [ 86%]
tests\metagpt\provider\test_ollama_api.py .                                                                                                                                 [ 87%]
tests\metagpt\provider\test_openai.py ........                                                                                                                              [ 93%]
tests\metagpt\provider\test_qianfan_api.py .                                                                                                                                [ 94%]
tests\metagpt\provider\test_spark_api.py ...                                                                                                                                [ 96%]
tests\metagpt\provider\test_zhipuai_api.py ..                                                                                                                               [ 98%]
tests\metagpt\provider\zhipuai\test_async_sse_client.py .                                                                                                                   [ 99%]
tests\metagpt\provider\zhipuai\test_zhipu_model_api.py .                                                                                                                    [100%]

======================================================================== 129 passed, 7 warnings in 10.59s ========================================================================
 ```

**completion test**

```python
models = ["cohere.command-text-v14", "mistral.mixtral-8x7b-instruct-v0:1", "meta.llama3-70b-instruct-v1:0",
              "anthropic.claude-3-haiku-20240307-v1:0", "ai21.j2-grande-instruct", "amazon.titan-text-express-v1"]
    for model_id in models:
        print(f"============{model_id}=============")
        my_config.model = model_id
        api = AmazonBedrockLLM(my_config)
        msg = [{"role": 'user', "content": "Please say: I am \{YOUR_MODEL_NAME\}"}]
        print(api.completion(msg))
        api._chat_completion_stream(msg)
```

```shell
2024-04-27 13:13:15.799 | INFO     | metagpt.const:get_metagpt_package_root:29 - Package root set to D:\python_projects\MetaGPT
============cohere.command-text-v14=============
 Hello, World!
 Hello, World!
============mistral.mixtral-8x7b-instruct-v0:1=============
 Hello World! I'm a helpful, respectful, and honest assistant, always ready to provide accurate and appropriate information. How can I assist you today? Is there a specific question or topic you'd like to discuss? I'm here to help make your life easier and more enjoyable.
 Hello World! I'm a helpful, respectful, and honest assistant, always ready to provide accurate and appropriate information. How can I assist you today? Is there a specific question or topic you'd like to discuss? I'm here to help make your life easier and more enjoyable.
============meta.llama3-70b-instruct-v1:0=============
Hello World!
Hello World!
============anthropic.claude-3-haiku-20240307-v1:0=============
Hello World!
Hello World!
============ai21.j2-grande-instruct=============
assistant: Hello World
============amazon.titan-text-express-v1=============
Sorry, but this model is not capable of saying "Hello World".
Sorry, but this model is not capable of saying "Hello World".
```

### CODE REVIEW

**Comment by geekan** [comment]
well done!

File: metagpt/provider/bedrock/amazon_bedrock_api.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+        self.__credentital_kwargs = {
+            "aws_secret_access_key": self.config.secret_key,
+            "aws_access_key_id": self.config.access_key,
+            "region_name": self.config.region_name
+        }
+        session = boto3.Session(**self.__credentital_kwargs)
+        client = session.client(service_name)
+        return client
+
+    def _get_client(self):
```
Comment (better629) [inline metagpt/provider/bedrock/amazon_bedrock_api.py:39]: _get_client use
@property
def client(self):
    return self._client?

same with _get_provider. but it seems that _get_client doesn't used in anyplace?

File: metagpt/provider/bedrock/amazon_bedrock_api.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+
+        ```shell
+        ai21.j2-ultra-v1                    Support Streaming:False
+        meta.llama3-70b-instruct-v1:0       Support Streaming:True
+        ……
+        ```
+        """
+        client = self.__init_client("bedrock")
+        # only output text-generation models
+        response = client.list_foundation_models(byOutputModality='TEXT')
```
Comment (better629) [inline metagpt/provider/bedrock/amazon_bedrock_api.py:56]: use "

File: metagpt/provider/bedrock/amazon_bedrock_api.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+        summaries = [f'{summary["modelId"]:50} Support Streaming:{summary["responseStreamingSupported"]}'
+                     for summary in response["modelSummaries"]]
+        logger.info("\n"+"\n".join(summaries))
+
+    def invoke_model(self, request_body) -> dict:
```
Comment (better629) [inline metagpt/provider/bedrock/amazon_bedrock_api.py:61]: add param type

File: metagpt/provider/bedrock/amazon_bedrock_api.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+        return response_body
+
+    def invoke_model_with_response_stream(self, request_body) -> EventStream:
+        response = self.__client.invoke_model_with_response_stream(
+            modelId=self.config.model, body=request_body
+        )
+        return response
+
+    @property
+    def _generate_kwargs(self) -> dict:
```
Comment (better629) [inline metagpt/provider/bedrock/amazon_bedrock_api.py:75]: _const_kwargs was used in other llms, keep same ?

File: metagpt/provider/bedrock/amazon_bedrock_api.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+    async def acompletion(self, messages: list[dict]):
+        return await self._achat_completion(messages)
+
+    async def acompletion_text(self, messages: list[dict], stream: bool = False,
+                               timeout: int = USE_CONFIG_TIMEOUT) -> str:
+        if stream:
+            return await self._achat_completion_stream(messages)
+        return await self._achat_completion(messages)
+
+    async def _achat_completion(self, messages: list[dict], timeout=USE_CONFIG_TIMEOUT):
```
Comment (better629) [inline metagpt/provider/bedrock/amazon_bedrock_api.py:128]: add output param type;

File: metagpt/provider/bedrock/amazon_bedrock_api.py [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+
+    # boto3 don't support support asynchronous calls.
+    # for asynchronous version of boto3, check out:
+    # https://aioboto3.readthedocs.io/en/latest/usage.html
+    # However,aioboto3 doesn't support invoke model
+
+    async def acompletion(self, messages: list[dict]):
+        return await self._achat_completion(messages)
+
+    async def acompletion_text(self, messages: list[dict], stream: bool = False,
```
Comment (better629) [inline metagpt/provider/bedrock/amazon_bedrock_api.py:122]: see acompletion_text in BaseLLM, there has a get_choice_text to extract final response text. Maybe you can put your self.__provider.get_choice_text inside it. So we can keep the whole structure.

File: metagpt/provider/bedrock/amazon_bedrock_api.py [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+            messages, self._generate_kwargs, stream=True)
+
+        response = self.invoke_model_with_response_stream(request_body)
+        collected_content = []
+        for event in response["body"]:
+            chunk_text = self.__provider.get_choice_text_from_stream(event)
+            collected_content.append(chunk_text)
+            log_llm_stream(chunk_text)
+
+        log_llm_stream("\n")
```
Comment (better629) [inline metagpt/provider/bedrock/amazon_bedrock_api.py:110]: miss _update_costs to calculate the llm api cost?

File: metagpt/provider/bedrock/amazon_bedrock_api.py [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+            max_tokens = model_max_tokens
+        else:
+            max_tokens = self.config.max_token
+
+        return {
+            self.__provider.max_tokens_field_name: max_tokens,
+            "temperature": self.config.temperature
+        }
+
+    def completion(self, messages: list[dict]) -> str:
```
Comment (better629) [inline metagpt/provider/bedrock/amazon_bedrock_api.py:87]: see other apis, it usually return a response structure not final extracted text.

File: metagpt/provider/bedrock/amazon_bedrock_api.py [hunk 9]
```diff
@@ -0,0 +1,136 @@
+from typing import Literal
```
Comment (better629) [inline metagpt/provider/bedrock/amazon_bedrock_api.py:1]: put the bedrock_api.py in metagpt/provider/bedrock_api.py? and keep others inside bedrock folder.

File: metagpt/provider/bedrock/bedrock_provider.py [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+
+    def get_request_body(self, messages: list[dict], generate_kwargs, *args, **kwargs):
+        body = json.dumps({
+            "inputText": self.messages_to_prompt(messages),
+            "textGenerationConfig": generate_kwargs
+        })
+        return body
+
+    def _get_completion_from_dict(self, rsp_dict: dict) -> str:
+        return rsp_dict['results'][0]['outputText']
```
Comment (better629) [inline metagpt/provider/bedrock/bedrock_provider.py:95]: use ", check other code.

File: tests/metagpt/provider/req_resp_const.py [hunk 11]
```diff
@@ -155,7 +160,8 @@ def get_anthropic_response(name: str, stream: bool = False) -> Message:
             ),
             ContentBlockDeltaEvent(
                 index=0,
-                delta=TextDelta(text=resp_cont_tmpl.format(name=name), type="text_delta"),
+                delta=TextDelta(text=resp_cont_tmpl.format(
```
Comment (better629) [inline tests/metagpt/provider/req_resp_const.py:163]: pre-commit you can set max_length of line to 120. 80 is too short. There is no need to auto change these code.

File: metagpt/provider/bedrock/amazon_bedrock_api.py [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+from metagpt.provider.base_llm import BaseLLM
+from metagpt.logs import log_llm_stream, logger
+from metagpt.provider.bedrock.bedrock_provider import get_provider
+from metagpt.provider.bedrock.utils import NOT_SUUPORT_STREAM_MODELS, get_max_tokens
+try:
+    import boto3
+    from botocore.eventstream import EventStream
+except ImportError:
+    raise ImportError(
+        "boto3 not found! please install it by `pip install boto3` ")
```
Comment (better629) [inline metagpt/provider/bedrock/amazon_bedrock_api.py:15]: if the independences and itself of boto3 is not huge, you can add it into requirements.txt.

File: metagpt/provider/bedrock/base_provider.py [hunk 13]
```diff
@@ -0,0 +1,29 @@
+import json
+from abc import ABC, abstractmethod
+
+
+class BaseBedrockProvider(ABC):
```
Comment (better629) [inline metagpt/provider/bedrock/base_provider.py:5]: any preferable name of `Provider` here to distinguish with folder name provider?

File: metagpt/provider/bedrock/amazon_bedrock_api.py (hunk 1 above)
Comment (usamimeri) [reply metagpt/provider/bedrock/amazon_bedrock_api.py:39]: It is used in test

File: metagpt/provider/bedrock/bedrock_provider.py (hunk 10 above)
Comment (usamimeri) [reply metagpt/provider/bedrock/bedrock_provider.py:95]: I encountered an issue earlier where I couldn't install pre-commit locally, so I've been using autopep8 for formatting but neglected this problem. I'll resolve this issue later.

File: metagpt/provider/bedrock/base_provider.py (hunk 13 above)
Comment (usamimeri) [reply metagpt/provider/bedrock/base_provider.py:5]: I currently don't have any ideas. Bedrock seems more like a transit station for other providers like claude or llama3.

File: metagpt/provider/bedrock/utils.py [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+    BOS, EOS = "<|begin_of_text|>", "<|eot_id|>"
+    GENERAL_TEMPLATE = "<|start_header_id|>{role}<|end_header_id|>\n\n{content}<|eot_id|>"
+
+    prompt = f"{BOS}"
+    for message in messages:
+        role = message["role"]
+        content = message["content"]
+        prompt += GENERAL_TEMPLATE.format(role=role, content=content)
+
+    if role != "assistant":
```
Comment (better629) [inline metagpt/provider/bedrock/utils.py:78]: role will have a NameError if messages is empty

File: metagpt/provider/bedrock/utils.py [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+        prompt += f"\n\nAssistant:"
+
+    return prompt
+
+
+def get_max_tokens(model_id) -> int:
+    return (NOT_SUUPORT_STREAM_MODELS | SUPPORT_STREAM_MODELS)[model_id]
```
Comment (better629) [inline metagpt/provider/bedrock/utils.py:99]: .get(model_id, default_val) to avoid KeyError

File: metagpt/provider/bedrock_api.py [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+            self.__provider.max_tokens_field_name: max_tokens,
+            "temperature": self.config.temperature
+        }
+
+    def completion(self, messages: list[dict]) -> str:
+        request_body = self.__provider.get_request_body(
+            messages, self._const_kwargs)
+        response_body = self.invoke_model(request_body)
+        completions = self.__provider.get_choice_text(response_body)
+        return completions
```
Comment (better629) [inline metagpt/provider/bedrock_api.py:93]: completion return response structure not text from get_choice_text.  see acompletion_text in BaseLLM

File: metagpt/provider/bedrock_api.py [hunk 17]
```diff
@@ -0,0 +1,141 @@
...
+        usage = self._get_usage(response)
+        self._update_costs(usage)
+        response_body = self._get_response_body(response)
+        return response_body
+
+    def invoke_model_with_response_stream(self, request_body: str) -> EventStream:
+        response = self.__client.invoke_model_with_response_stream(
+            modelId=self.config.model, body=request_body)
+        usage = self._get_usage(response)
+        self._update_costs(usage)
```
Comment (better629) [inline metagpt/provider/bedrock_api.py:72]: Here doesn't have the price of each 1k tokens for model, so `_update_costs` will fail. Don't you notice this in your log?

File: metagpt/provider/bedrock_api.py [hunk 18]
```diff
@@ -0,0 +1,141 @@
...
+            messages, self._const_kwargs)
+        response_body = self.invoke_model(request_body)
+        completions = self.__provider.get_choice_text(response_body)
+        return completions
+
+    def _chat_completion_stream(self, messages: list[dict], timeout=USE_CONFIG_TIMEOUT) -> str:
+        if self.config.model in NOT_SUUPORT_STREAM_MODELS:
+            logger.warning(
+                f"model {self.config.model} doesn't support streaming output!")
+            return self.completion(messages)
```
Comment (better629) [inline metagpt/provider/bedrock_api.py:99]: notice here, completion return object not str, see comment like before.

**Review by geekan — APPROVED** [review]
LGTM

**Pull request merged**